### PR TITLE
Simplify lakeformation shares - single shared_db and single resource link table

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Feature

### Detail
Implements design specified in #746. Multiple share requests share a single shared Glue database instead of a shared database per share request.  Same for resource link tables, multiple shares in the same target account get permissions to the same resource link table.

In addition, this PR includes:
- Unified LakeFormation processor for same account and cross account shares (remove duplicated code)
- Code clean-up of Lake Formation manager and client: remove duplicates, added more error handling
- Ensure "no-failures" if a resource already exists or is already deleted or if a permission is already granted or revoked
- Redesigned tests and increased coverage

#### Upgrade LakeFormation version
LakeFormation v3 ([docs](https://docs.aws.amazon.com/lake-formation/latest/dg/optimize-ram.html)) simplifies cross-account sharing requests. We want to see if there are advantages of implementing the capability of LakeFormation to define the external IAM role directly, instead of using first an AWS account and then the role. In addition, v4 introduces LF hybrid mode. We are not using it at the moment, but it can be a great addition in future features. Because we already distinguish between "old" and "new" shares, it is relatively easy to upgrade to a new LF version without affecting "old" shares.

**Does it make sense to upgrade the sharing mechanism?**
- Is it possible technically for all principals? --> No, which is a blocker
```Could not grant principal arn:aws:quicksight:us-east-1:11111111111:group/default/dataall permissions ['DESCRIBE', 'SELECT'] to {'Table': {'DatabaseName': 'dataall_imported_c3_tvnews81', 'Name': 'children_books_raw', 'CatalogId': '2222222222222'}}  due to: An error occurred (InvalidInputException) when calling the GrantPermissions operation: Cross account requests are only allowed for AWS Accounts, Organizations, IAM Principals and All IAMPrincipals```

- What actions are simplified? In this [commit ](https://github.com/data-dot-all/dataall/pull/1016/commits/f5b9bee5e62683859959392e98c094e25df5fb5e) we can see how the final implementation would look like. 

**How does the migration look like?**

1. Upgrade LF version: It is possible to change the LF cross account settings programmatically using [put_data_lake_settings](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lakeformation/client/put_data_lake_settings.html) API. We already use this API in a custom resource of the environment stack. We could introduce the LF version as part of that call and make sure that it is executed on update (not only on create). 
2. Modify code in sharing task: for old_shares use the AWS account and for new shares use the IAM roles as principals of the grant. for old_shares execute additional grant for principals in target, for new shares skip. See [commit ](https://github.com/data-dot-all/dataall/pull/1016/commits/f5b9bee5e62683859959392e98c094e25df5fb5e).
3. Backwards compatibility for shares done with V2:
  - At the moment the code is meant to work for versions >= v2. v4 is very recent and most AWS accounts use v3
  - Grant: If a RAM request is v2 and the LF version is upgraded to v3, new tables added to the RAM invitation can use v3 underlying code (I tested one table shared with old v2 procedure and added one table to the request with new code changes and it succeeds. In source the principals are the target IAM roles not the target account)
  - Revoke: not tested

**Conclusion**
- We cannot implement LF v3 direct sharing with principals because it does not support Quicksight groups.
- I think that setting the LF version in the environment stack is a good idea regardless of it being explicitly used It gives us control to set the latest version in the accounts we use. This is a feature that should be part of a separate PR.
- v2 RAM invitations still work with v3 direct-IAM role code. Therefore we could keep using v2 shares without the need to inspect shares and upgrade them. This means that the upgrade is pretty seamless, and we can take it at a later time when the Quicksight blocker is resolved.

Issues resolved:
- [X] Issue: deletion of new "shared" database does not happen when other old shares are still present in environment account.
- [X] ~[Pre-existing issue]:  In cross-account shares revoke ALL does not revoke permissions to QS group (by-design)~
- [X]  [Pre-existing issue]: in same account requesters have access to original table. It is needed! If we remove access to the original table the resource link access to data does not work.
- [X] ~Issue: when revoking, permissions to the shared db are not revoked for principal~
- [X] ~Issue: shared db is not deleted when all shares are revoked~
- [X] ~Issue: duplication of LF permissions. Pivot role and QS group were granted permissions on the original db and on the shared db with every new share~
- [ ]  [Pre-existing issue]: orphan LF resources on failed shares. If a share fails it can be deleted from data.all UI but there are resources that are created in LF that require manual clean-up (ugly operation assuming pivot role) 
- [X]  ~Changes in UI: data consumption details~

### Next steps and remarks
- Because there are shared resources between shares, it is possible that parallel shares on the same tables have conflicts: e.g. 2 shares trying to create a resource link simultaneously. I have tried testing with shares approved in a matter of seconds and did not run into issues (at the end a look-up is way faster than updating a policy). But we should keep an eye on this.
- Upgrade Lake Formation version. Not possible at this moment, because it would not work for Quicksight groups. To be re-evaluated in the future.
- I think the orphan LF resources on failed shares can be done in a separate PR to limit the scope of this PR. 
- I tried to revoke the requester access to the original table in same account sharing, but it is not possible. For the requester to be able to access the underlying data of the resource link, it needs to have access to the table. 
- The issue "deletion of new "shared" database does not happen when other old shares are still present in environment account." has not a big impact, but I implemented a solution nevertheless.

### Relates
- #746 

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dlpzx** [comment]
## Testing in AWS
- [x] CICD pipeline completes successfully

### Same Account
- [ ] New sharing
  - process share1 with 1 table - table queried using Athena
  - process share2 for 2 tables - tables queried using Athena, only one additional tables, no additional databases created. And no duplication of permissions (e.g. permissions to pivot role in shared-db)
  - process share3 for the same 2 tables (with consumption role that is a service role)
  - Revoke share3 items - removes ALL permissions to principal 3 but tables and permissions for shares 1 and 2 still work. Permissions to pivot role to shared and original db remain
  - Revoke share2 items - removes permissions to principal 2, deletes one of the resource link tables, permissions and share 1 still work. Permissions to pivot role to shared and original db remain
  - Revoke share1 items- cleans-up; deletes permissions from source and deletes shared db and resource link table



- [ ] Backwards compatibility. Create 2 shares with old version of the code. Share1 with one table, share2 with 2 tables. Upgrade the code.
  - Add table to share1 - new resource link to the "shared-uri1" database. No "shared" database created. Both tables queriable in Athena.
  - Create brand new share (share3) on the same dataset - creates "shared" database and resource link for table. share 1 and share 2 unaffected. No new RAMs
  - Revoke one table from share2 - deletes resource link from "shared-uri2" database. share 2 for the remaining table still has permissions and can be queried in Athena. share1 and share3 unaffected.
  - Revoke all items from share2 - deletes "shared-uri2" database. share1 and share3 unaffected
  - Revoke all items from share3 - deletes "shared" database. share1 unaffected
  - Re-create share2 - creates "shared" database and resource links in this database

### Cross Account
- [ ] New sharing
  - process share1 with 1 table - table queried using Athena. RAM invitation with database and table in the share
  -  process share2 for 2 tables - tables queried using Athena, only one additional tables, no additional databases created. And no duplication of permissions (e.g. permissions to pivot role in shared-db). Single RAM invitation for both shares containing 2 tables.
  - Revoke share2 items - removes permissions to principal 2, deletes one of the resource link tables, permissions and share 1 still work. Permissions to pivot role to shared and original db remain. In RAM invitation the non-shared table is "Disassociated". The rest of the items still shared.
  - Revoke share1 items- cleans-up; deletes permissions from source and deletes shared db and resource link table 


- [ ] Backwards compatibility. Create 2 shares with old version of the code. Share1 with one table, share2 with 2 tables. Upgrade the code.
  - Add table to share1 - new resource link to the "shared-uri1" database. No "shared" database created. Both tables queriable in Athena. UI shows correct database name.
  - Create brand new share (share3) on the same dataset - creates "shared" database and resource link for table. share 1 and share 2 unaffected. No new RAMs 
  - Revoke one table from share2 - deletes resource link from "shared-uri2" database. share 2 for the remaining table still has permissions and can be queried in Athena. share1 and share3 unaffected. 
  - Revoke all items from share2 - deletes "shared-uri2" database. share1 and share3 unaffected
  - Revoke all items from share3 - deletes "shared" database. share1 unaffected 
  - Revoke one item from share1 - deletes permissions for that item, also QS permissions

File: backend/dataall/modules/dataset_sharing/aws/glue_client.py [hunk 1]
```diff
@@ -39,89 +49,141 @@ def _create_glue_database(self, location):
...
+            )
+            return response
+        except ClientError as e:
+            log.error(
+                f'Could not delete table {table_name} '
+                f'in database {database} '
+                f'in catalog {self._account_id} '
+                f'due to: {e}'
+            )
+            raise e
```
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/aws/glue_client.py:126]: `delete_table` is used in `dataall.modules.dataset_sharing.services.share_managers.lf_share_manager.LFShareManager.delete_resource_link_table_in_shared_database` and you are right, the exception is not raised there. Instead, the exception is catched directly in `dataall.modules.dataset_sharing.services.share_processors.lakeformation_process_share.ProcessLakeFormationShare.process_revoked_shares`

[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/aws/glue_client.py:126]: It looks a bit scary, so I think we should test some "failure" scenarios. In the validation we could include some scenarios like: process share without source table (which should result in an error and the sharing going to failure). During the development I faced failure scenarios so I was able to test the try/except, but it does not harm to validate

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                resource_link_table_exists = self.check_resource_link_table_exists_in_target_database(table)
+                other_table_shares_in_env = True if ShareObjectRepository.other_approved_share_item_table_exists(
+                    self.session,
+                    self.target_environment.environmentUri,
+                    share_item.itemUri,
+                    share_item.shareItemUri
+                ) else False
+
+                if resource_link_table_exists:
+                    self.revoke_principals_permissions_to_resource_link_table(table)
```
Comment (noah-paige) [inline backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:180]: do we not need to check `other_table_shares_in_env` to determine if we need to remove QS principal (if dashboards enabled) before revoking here as well?

**Comment by noah-paige** [comment]
Overall - this PR looks great! Found 1 possible error on handling QS Permissions but I think there is a straight forward fix, outcomes of my testing are below:

### Cross Account
Backwards compatibility
- [X] Create Share share1 with one table shared. Upgrade the code
- [X] UI shows correct shared_db name
- [X] Add table to share1 - new resource link to the "shared-uri1" database. No "shared" database created. Both tables queriable in Athena
- [X] Create brand new share (share2) on the same dataset - creates "shared" database and resource link for table. share 1 unaffected
- [X] Revoke one table from share1 - deletes resource link from "shared-uri1" database. share1 for the remaining table still has permissions and can be queried in Athena. share2 unaffected
- [X] Revoke all items from share2 - deletes "shared" database. share1 unaffected
- [X] Re-create share2 and Revoke all items from share1 - deletes "shared-uri1" database. share1 unaffected

New Sharing
- [X] process share1 with 1 tables - table queried using Athena. RAM invitation with database and table in the share
- [X] process share2 for 2 tables - tables queried using Athena, only one additional tables, no additional databases created. And no duplication of permissions (e.g. permissions to pivot role in shared-db). Single RAM invitation for both shares containing 2 tables
- [X] Create QS Dashboard from table shared in share1 
- [ ] Revoke share2 items - removes permissions to principal 2, deletes one of the resource link tables, permissions and share 1 still work. Permissions to pivot role & QS service role to shared and original db and remain. In RAM invitation the non-shared table is "Disassociated". The rest of the items still shared and QS Dashboard still accessible
 —> ERROR: QS Permissions are removed at this step and QS Dashboard gets Insufficient LF Permission Error (no longer accessible)
- [X] Revoke share1 items- cleans-up; deletes permissions from source and deletes shared db and resource link table

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py (hunk 2 above)
Comment (noah-paige) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:180]: Found a potential error with removing QS Permissions - check out testing outcomes for more detail

File: backend/dataall/modules/dataset_sharing/db/share_object_repositories.py [hunk 3]
```diff
@@ -1026,7 +1027,7 @@ def paginated_dataset_shares(session, uri, data=None) -> [ShareObject]:
         ).to_dict()
 
     @staticmethod
-    def list_dataset_shares_with_existing_shared_items(session, dataset_uri) -> [ShareObject]:
+    def list_dataset_shares_with_existing_shared_items(session, dataset_uri, environment_uri=None, item_type=None) -> [ShareObject]:
```
[the remark this answers is not in the document]
Comment (noah-paige) [reply backend/dataall/modules/dataset_sharing/db/share_object_repositories.py:1030]: Or is this just a placeholder for v2.6 when old share methods are considered deprecated?

[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/db/share_object_repositories.py:1030]: Placeholder for v2.6, at that moment we will need the environment_uri

File: backend/dataall/base/utils/alarm_service.py
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/base/utils/alarm_service.py:1]: You mean that the `AlarmService.publish_message_to_alarms_topic` does not return anything and thus we cannot verify its output right? Although in general I agree that we should always validate function output, in this case I don't think it makes much difference. `AlarmService.publish_message_to_alarms_topic` is a function triggered by an exception. Some error in the code >>>> error handling in backend >>> publish error to SNS. If there is an error in publish error to SNS I am not sure what actions to take. In other words, what should be the error handling of the error handling?

[the remark this answers is not in the document]
Comment (SofiaSazonova) [reply backend/dataall/base/utils/alarm_service.py:1]: Hm...yes, a bit too deep, I agree.

File: backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+                    session=self.session,
+                    uri=self.share.shareUri,
+                    item_type=ShareableType.Table.value
+                )
+                log.info(
+                    f'Remaining tables shared in this share object = {existing_shared_tables_in_share}')
+
+                if not existing_shared_tables_in_share:
+                    log.info("Revoking permissions to target shared database...")
+                    self.revoke_principals_database_permissions_to_shared_database()
```
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:212]: You are right, the permissions need to remain for the database, I'll fix it. The reason that the dashboard still can see the data is that for the table it only needs select permissions. I think the problem would be creating another Quicksight dataset for that table, in that case I doubt the Glue database appears in the Quicksight UI

[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:212]: Your comment made me think, do we need to remove quicksight permissions at all? We need to remove Quickdight permissions at the same time that we need to delete the table or the database. When deleting the object, the permissions are also deleted, so we could always skip the revoke for QS group right?. What do you think?

[the remark this answers is not in the document]
Comment (noah-paige) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:212]: with the way we have it currently designed I think that is correct - we technically only need to grant QS permissions on first time table is shared with target account and revoke only when delete of resource linked table

[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:212]: In this [commit](https://github.com/data-dot-all/dataall/commit/2aeecd74886b927f73d3c6a0af78a97c0de313b9) I have set the default to not revoke permissions to Quicksight groups. I have not tested it though

[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:212]: Tested in AWS:
-[x] CICD pipeline completes
- [ ] share1: Share a dataset with group in an environment with Dashboards enabled 
- [ ] share2: Share the same dataset with group2 in the same environment as share1
- [ ] revoke share2, QS group still has access to data
- [ ] revoke share1, QS permissions removed. No access to data from dashboard

[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/dataset_sharing/services/share_processors/lakeformation_process_share.py:212]: Re-tested after latest [commit](https://github.com/data-dot-all/dataall/commit/4af9d6d2b173eb688776fa98780d6f9fa284bd1c):
- [X] CICD pipeline completes
- [x] share1: Share a dataset (table1) with group in an environment with Dashboards enabled - create Dashboard accessing table1 data
- [X] share2: Share the same dataset (table1 and table2) with group2 in the same environment as share1 - create Dashboard accessing table2 data
- [X] revoke share2, QS group still has access to table1 data, table2 data dashboard cannot refresh new data (only spice data) --> dashboard with table1 still sees data
- [X] revoke share1, QS permissions removed. No access to data from dashboard
All permissions in Lake Formation are cleaned when it is required

**Comment by noah-paige** [comment]
Reviewed changes and looks good! Testing once more now for table share(s) approve/reject

**Pull request merged**